Replicator/Replicator.cc

```diff
@@ -88,66 +87,19 @@ namespace litecore { namespace repl {
-                if (c == nullptr) {
-                    error::_throw(error::UnexpectedError);
-                }
-                this->_subRepls[i].collection = c;
-            });
-        });
-
-        actor::Timer::duration saveDelay = tuning::kDefaultCheckpointSaveDelay;
-        if (auto i = options->properties[kC4ReplicatorCheckpointInterval].asInt(); i > 0)
-            saveDelay = chrono::seconds(i);
-
-        bool isPushBusy = false;
-        bool isPullBusy = false;
-        for (CollectionIndex i = 0; i < _options->collectionCount(); ++i) {
-            SubReplicator& sub = _subRepls[i];
-            if (_options->push(i) != kC4Disabled) {
-                sub.checkpointer.reset(new Checkpointer(_options, webSocket->url(), sub.collection));
-                sub.pusher = new Pusher(this, *sub.checkpointer, i);
-                sub.pushStatus = kC4Busy;
-                isPushBusy = true;
-            } else {
-                sub.pushStatus = kC4Stopped;
-            }
-            if (_options->pull(i) != kC4Disabled) {
-                sub.puller = new Puller(this, i);
-                sub.pullStatus = kC4Busy;
-                if (sub.checkpointer.get() == nullptr) {
-                    sub.checkpointer.reset(new Checkpointer(_options, webSocket->url(), sub.collection));
-                }
-                isPullBusy = true;
-            } else {
-                sub.pullStatus = kC4Stopped;
-            }
-            DebugAssert(sub.checkpointer.get() != nullptr);
-            sub.checkpointer->enableAutosave(saveDelay, bind(&Replicator::saveCheckpoint, this, i, _1));
+        _remoteURL = webSocket->url();
+        if (_options->isActive()) {
+            prepareWorkers();
         }
-        _pushStatus = isPushBusy ? kC4Busy : kC4Stopped;
-        _pullStatus = isPullBusy ? kC4Busy : kC4Stopped;
 
-        if (std::none_of(_subRepls.begin(), _subRepls.end(), [](const SubReplicator& sub) {
-            return sub.pusher;
-        })) {
-            for (auto profile : {"subChanges", "getAttachment", "proveAttachment"})
-                registerHandler(profile,  &Replicator::returnForbidden);
-        }
-        if (std::none_of(_subRepls.begin(), _subRepls.end(), [](const SubReplicator& sub) {
-            return sub.puller;
-        })) {
-            for (auto profile : {"changes", "proposeChanges", "rev", "norev"})
-                registerHandler(profile,  &Replicator::returnForbidden);
+        // Following messages are handled by appropriate workers.
+        // Replicator receives all the messages. Based on collectionIndex,
+        // it dispatches the message to appropriate workers.
+        for (auto profile : {"subChanges", "getAttachment", "proveAttachment", // passive pushers
+                "changes", "proposeChanges", "rev", "norev"                    // passive pullers
+        }) {
+            registerHandler(profile, &Replicator::handleConnectionMessage);
```

Comment:
I think this can be done in registerWorkerHandler() when registering the profile first time.

Reply:
I understand that but I understand that registering the BLIP message handler can be done in registerWorkerHandler() as well so there is no need to do it separately upfront.

Reply:
We may have 5 pushers. Each pusher will register as a worker handler for, say, "change", message. But for Replicator, we only register once with BLIPIO